File: go/pkg/embed/dashscope.go
package embed

import (
	"context"
	"fmt"
	"net/http"

	"github.com/openai/openai-go"
	"github.com/openai/openai-go/option"
)

// DashScope embedding models.
const (
	// ModelDashScopeV4 is the latest DashScope embedding model.
	// Supports 100+ languages, dimensions: 64–2048, default 1024.
	ModelDashScopeV4 = "text-embedding-v4"

	// ModelDashScopeV3 supports 50+ languages, dimensions: 64–1024.
	ModelDashScopeV3 = "text-embedding-v3"

	// ModelDashScopeV2 has fixed 1536 dimensions.
	ModelDashScopeV2 = "text-embedding-v2"

	// ModelDashScopeV1 has fixed 1536 dimensions.
	ModelDashScopeV1 = "text-embedding-v1"
)

const (
	dashScopeBaseURL      = "https://dashscope.aliyuncs.com/compatible-mode/v1"
	dashScopeMaxBatch     = 10 // v3/v4 max batch size
	dashScopeDefaultDim   = 1024
	dashScopeDefaultModel = ModelDashScopeV4
)

// DashScope implements [Embedder] using Aliyun DashScope's OpenAI-compatible
// embedding API.
type DashScope struct {
	client *openai.Client
	model  string
	dim    int
}

var _ Embedder = (*DashScope)(nil)

// NewDashScope creates a DashScope embedder.
//
// The apiKey is required and can be obtained from:
// https://bailian.console.aliyun.com/?apiKey=1
func NewDashScope(apiKey string, opts ...Option) *DashScope {
	cfg := config{
		model:      dashScopeDefaultModel,
		dim:        dashScopeDefaultDim,
		baseURL:    dashScopeBaseURL,
		httpClient: http.DefaultClient,
	}
	for _, o := range opts {
		o(&cfg)
	}

	client := openai.NewClient(
		option.WithAPIKey(apiKey),
		option.WithBaseURL(cfg.baseURL),
		option.WithHTTPClient(cfg.httpClient),
	)

	return &DashScope{
		client: &client,
		model:  cfg.model,
		dim:    cfg.dim,
	}
}

// Embed returns the embedding for a single text.
func (d *DashScope) Embed(ctx context.Context, text string) ([]float32, error) {
	if text == "" {
		return nil, ErrEmptyInput
	}
	vecs, err := d.EmbedBatch(ctx, []string{text})
	if err != nil {
		return nil, err
	}
	return vecs[0], nil
}

// EmbedBatch returns embeddings for multiple texts.
// Batches larger than 10 are automatically split into multiple API calls.
func (d *DashScope) EmbedBatch(ctx context.Context, texts []string) ([][]float32, error) {
	if len(texts) == 0 {
		return nil, ErrEmptyInput
	}

	result := make([][]float32, len(texts))
	for i := 0; i < len(texts); i += dashScopeMaxBatch {
		end := min(i+dashScopeMaxBatch, len(texts))
		batch := texts[i:end]

		vecs, err := d.callAPI(ctx, batch)
		if err != nil {
			return nil, fmt.Errorf("embed batch [%d:%d]: %w", i, end, err)
		}
		copy(result[i:], vecs)
	}
	return result, nil
}

// Dimension returns the configured vector dimensionality.
func (d *DashScope) Dimension() int {
	return d.dim
}

// Model returns the DashScope model identifier (e.g., "text-embedding-v4").
func (d *DashScope) Model() string {
	return d.model
}

func (d *DashScope) callAPI(ctx context.Context, texts []string) ([][]float32, error) {
	params := openai.EmbeddingNewParams{
		Model:          d.model,
		Input:          openai.EmbeddingNewParamsInputUnion{OfArrayOfStrings: texts},
		Dimensions:     openai.Int(int64(d.dim)),
		EncodingFormat: openai.EmbeddingNewParamsEncodingFormatFloat,
	}

	resp, err := d.client.Embeddings.New(ctx, params)
	if err != nil {
		return nil, err
	}

	vecs := make([][]float32, len(texts))
	for _, item := range resp.Data {
		idx := item.Index
		if idx < 0 || idx >= int64(len(texts)) {
			return nil, fmt.Errorf("unexpected embedding index %d for batch size %d", idx, len(texts))
		}
		vecs[idx] = float64sToFloat32s(item.Embedding)
	}

	// Verify all slots are filled.
	for i, v := range vecs {
		if v == nil {
			return nil, fmt.Errorf("missing embedding for index %d", i)
		}
	}
	return vecs, nil
}

// float64sToFloat32s converts a []float64 to []float32.
func float64sToFloat32s(f64 []float64) []float32 {
	f32 := make([]float32, len(f64))
	for i, v := range f64 {
		f32[i] = float32(v)
	}
	return f32
}

File: rust/embed/src/dashscope.rs
use reqwest::Client;

use crate::config::EmbedConfig;
use crate::embed::Embedder;
use crate::error::EmbedError;

/// DashScope embedding models.
pub const MODEL_DASHSCOPE_V4: &str = "text-embedding-v4";
pub const MODEL_DASHSCOPE_V3: &str = "text-embedding-v3";
pub const MODEL_DASHSCOPE_V2: &str = "text-embedding-v2";
pub const MODEL_DASHSCOPE_V1: &str = "text-embedding-v1";

const DASHSCOPE_BASE_URL: &str = "https://dashscope.aliyuncs.com/compatible-mode/v1";
const DASHSCOPE_MAX_BATCH: usize = 10;
const DASHSCOPE_DEFAULT_DIM: usize = 1024;

/// DashScope embedder using Aliyun DashScope's OpenAI-compatible API.
pub struct DashScope {
    client: Client,
    api_key: String,
    model: String,
    dim: usize,
    base_url: String,
}

impl DashScope {
    pub fn new(api_key: &str) -> Self {
        Self {
            client: Client::new(),
            api_key: api_key.to_string(),
            model: MODEL_DASHSCOPE_V4.to_string(),
            dim: DASHSCOPE_DEFAULT_DIM,
            base_url: DASHSCOPE_BASE_URL.to_string(),
        }
    }

    pub fn with_config(api_key: &str, cfg: EmbedConfig) -> Self {
        Self {
            client: Client::new(),
            api_key: api_key.to_string(),
            model: if cfg.model.is_empty() {
                MODEL_DASHSCOPE_V4.to_string()
            } else {
                cfg.model
            },
            dim: if cfg.dimension == 0 {
                DASHSCOPE_DEFAULT_DIM
            } else {
                cfg.dimension
            },
            base_url: if cfg.base_url.is_empty() {
                DASHSCOPE_BASE_URL.to_string()
            } else {
                cfg.base_url
            },
        }
    }

    async fn call_api(&self, texts: &[&str]) -> Result<Vec<Vec<f32>>, EmbedError> {
        crate::openai_compat::call_embedding_api(
            &self.client,
            &self.api_key,
            &self.base_url,
            &self.model,
            self.dim,
            texts,
        )
        .await
    }
}

#[async_trait::async_trait]
impl Embedder for DashScope {
    async fn embed(&self, text: &str) -> Result<Vec<f32>, EmbedError> {
        if text.is_empty() {
            return Err(EmbedError::EmptyInput);
        }
        let vecs = self.embed_batch(&[text]).await?;
        Ok(vecs.into_iter().next().unwrap())
    }

    async fn embed_batch(&self, texts: &[&str]) -> Result<Vec<Vec<f32>>, EmbedError> {
        if texts.is_empty() {
            return Err(EmbedError::EmptyInput);
        }

        let mut result = Vec::with_capacity(texts.len());
        for chunk in texts.chunks(DASHSCOPE_MAX_BATCH) {
            let vecs = self.call_api(chunk).await?;
            result.extend(vecs);
        }
        Ok(result)
    }

    fn dimension(&self) -> usize {
        self.dim
    }
}
